src/network/networkd-link.c

```diff
@@ -1717,6 +1717,22 @@ bool link_has_carrier(Link *link) {
         return false;
 }
 
+bool link_has_foreign_addresses(Link *link) {
+        Address *address;
+        Iterator i;
+
+        SET_FOREACH(address, link->addresses_foreign, i) {
+                /* we consider IPv6LL addresses to be managed by the kernel */
+                if (address->family == AF_INET6 && in_addr_is_link_local(AF_INET6, &address->in_addr) == 1)
+                        continue;
+
+                return true;
+        }
+
+        return false;
```

Comment:
Hmm, why this logic? We generally say that networkd either manages a link or it doesn't, but that managing a link with foreign IP addresses is explicitly not something we want to support. And I'd claim that a link that is configured to be kept in "down" state also falls under "managed", hence I think this function shouldn't be there. 

Can you elaborate?

Reply:
I got to this in an attempt to not only shutdown the link if there isn't already an address on it; for such cases as if you're restarting systemd-networkd for whatever reason. My thought was to not shutdown the link if it was already set to up previously, so as not to cause a network to go down just because systemd-networkd was restarted. Normally at boot, there is no address except for the link-local IPv6 address the kernel might have set.

Reply:
To be clear, this is meant to keep the network link down, up until the admin runs 'ip link set X up' to force it up -- at that point systemd-network should finish configuring as appropriate, get a DHCP address, etc.

Reply:
> To be clear, this is meant to keep the network link down, up until the admin runs 'ip link set X up' to force it up -- at that point systemd-network should finish configuring as appropriate, get a DHCP address, etc.

But that's very different from how we currently handle stuff. We assume that networkd is entirely in control of an interface and nobody else is. Such a "hybrid control" thing would be new territory. I am not sure I like the prospect of that... I mean, with that it becomes very unclear what a networkd restart should do: should it actually down the iface again or not?